OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs

```diff
@@ -141,6 +141,46 @@ public int ResourceDensityAt(CPos c)
 			return Math.Max(int2.Lerp(0, type.Info.MaxDensity, adjacent, 9), 1);
 		}
 
+		IEnumerable<CPos> NonEmptyNeighbours(CPos cell)
+		{
+			var neighbours = CVec.Directions.Select(d => d + cell);
+			foreach (var neighbour in neighbours)
+			{
+				if (Tiles[neighbour].Type != null)
+					yield return neighbour;
+			}
+		}
+
+		IEnumerable<CPos> ExpandNeighboursAround(CPos cell)
+		{
+			var patch = NonEmptyNeighbours(cell).Append(cell);
+			var count = 0;
+
+			do
+			{
+				count = patch.Count();
+				patch = patch.SelectMany(c => NonEmptyNeighbours(c)).Distinct();
+			} while (patch.Count() > count);
+
+			return patch;
+		}
+
+		public int GetValueFromPatchAround(CPos startCell)
```

Comment:
This is extremely slow, taking around 35ms(!) to calculate the value when moving the mouse across the central ore patch on the RA map "Barracuda".  To make the performance acceptable the `EditorResourceLayer` will need to maintain a `CellLayer<int>` connectivity index (like `DomainIndex` is for movement) and a `Dictionary<int, int>` to store the value of each connected field.  The field value can then be updated incrementally on each field edit, instead of on mouse movement.

Reply:
Sounds a bit complicated and I am not sure if I understood it correctly:

```csharp
int resourcePatches; // increment by one when a new patch is identified
CellLayer<int> resourcePatch; // store the resource patch number per cell
Dictionary<int, int> resourcePatchValue; // lookup the pre-calculated value per resource patch
```